Fast initial sync

mpetrunic opened the pull request:
resolves #479

mpetrunic committed:
add util method to rep store

ec2 committed:
move state transition into new package

ec2 committed:
add sorting function to eth2.0-utils

ec2 committed:
copied constants folder from lodestar package

ec2 committed:
moved all the state transition stuff to new package

ec2 committed:
change all state transition imports to point to newly created package

ec2 committed:
update package.json

ec2 committed:
update yarn lock

mpetrunic committed:
simple round robin array

mpetrunic committed:
sync utils

mpetrunic committed:
initial sync algorithm

mpetrunic committed:
add more comments

mpetrunic committed:
add sync utils tests

wemeetagain commented on packages/lodestar/src/sync/initial.ts:
```diff
@@ -5,86 +5,106 @@
+      chunks = (await Promise.all(
+        chunks.map(async (chunk) => {
+          try {
+            const chunkBlocks = await this.syncChunk(peers.next(), chunk);
```
does this request chunks in parallel? Or is it blocking on each request to finish?

mpetrunic replied:
in parallel, map will return array of promises to resolve

wemeetagain commented on packages/lodestar/src/sync/initial.ts:
```diff
@@ -5,86 +5,106 @@
+    }
+  };
+
+  private async syncToTarget(targetEpoch: Epoch): Promise<BeaconBlock[]> {
```
maybe call this batchRequestBlocks, or something along those lines, we aren't really syncing here, just fetching blocks.

mpetrunic replied:
yeah already renamed it

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,39 @@
+  return currentCheckPoint.epoch;
+}
+
+export type Chunk = [Slot, Slot];
```
Maybe call this BlockRange?

wemeetagain reviewed (commented):
I feel like we should be able to start processing blocks as we receive them, rather than saving them all up until the end.
Also I wonder what the tradeoffs with using the `step` param in beacon_block to interleave requests look like, rather than requesting contiguous ranges.

GregTheGreek commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,39 @@
+
+export function getTargetEpoch(peers: IReputation[], currentCheckPoint: Checkpoint): Epoch {
+  const peersWithHigherFinalizedEpoch = peers.filter(peer => {
+    if(!peer.latestHello) {
```
How recent are the hello's?

mpetrunic replied:
not recent, I don't think we are refreshing hellos, but I would do this in another PR as it's needed in regular sync as well. Possible solution, every SECONDS_PER_SLOT go trough all peers and refresh hellos.

GregTheGreek replied:
Yep!

Intitially i was thinking they would be new because of startup, but the more i thought about it, we could lag at any point need to resync and our hello's our totally out of wack

Another thing to think about is when we actually start syncing. Do we have an ideal `PEER_THRESHOLD` where we hold off on syncing until we have an ideal amount of peers

mpetrunic replied:
I put 1 for now, as we will test it against single node most of the time. But I will add that as cli option in this PR so user can configure (I think you can do this in current eth1 nodes).

I left peers as reference to peers in sync module so sync can refresh peer data (and it will reflect in regular and initial sync) without some complicated passing on.

mpetrunic committed:
initial sync algorithm

mpetrunic committed:
improve gitignore

mpetrunic committed:
add processedEpoch chain event

mpetrunic committed:
sync util tests

Commit:
Convert interop command to dev command

mpetrunic committed:
block fetching tests

mpetrunic committed:
initial sync tests

mpetrunic committed:
fix tests

mpetrunic committed:
lint fix

mpetrunic committed:
sort blocks when validating

ec2 committed:
fix state transition unit tests

ec2 committed:
fix lodestar tests

Commit:
Successfully run ./bin/lodestar dev

Commit:
Fix linting

ec2 committed:
export everything in epoch

ec2 committed:
fix e2e tests

ec2 committed:
fix spec tests

ec2 committed:
remove unused import

ec2 committed:
add bls dependancy

ec2 committed:
linting

ec2 committed:
update yarn lock

wemeetagain commented on packages/lodestar/src/chain/chain.ts:
```diff
@@ -338,6 +339,7 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
     // Post-epoch processing
     const currentEpoch = computeEpochOfSlot(this.config, newState.slot);
     if (computeEpochOfSlot(this.config, preSlot) < currentEpoch) {
+      this.emit("processedEpoch", {epoch: currentEpoch, root: blockRoot});
```
maybe `processedCheckpoint` to match the other events

mpetrunic replied:
sure. 

Side note: we have quite a lot of different event name styles. Do we wan't at some point to have processed:checkpoint or processedCheckpoint

wemeetagain replied:
probably worth making an issue to straighten out the event names.
Maybe `chain:processedCheckpoint` would eventually be the best here.

wemeetagain commented on packages/lodestar/src/sync/initial/initial.ts:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    const chainCheckPoint = this.chain.latestState.currentJustifiedCheckpoint;
+    //listen on finalization events
```
We also have a finalizedCheckpoint event. Is that the finalization event we want here?

mpetrunic replied:
you would think so. But you will receive finalizedCheckpoint event only after you justify at least finalizedEpoch+1 which will never happen in initial sync

ec2 replied:
That's confusing 😕 ...

mpetrunic replied:
think about finalized is really really justified epoch 😄

ec2 committed:
fix paths

ec2 committed:
removed network constants resulted from copy pasta

ec2 committed:
use older bit-utils library

ec2 committed:
linting fixes

ec2 committed:
fix finalUpdates test

ColinSchwarz committed:
Update README.md

not sure if  this is the right way to do a link without having the whole URL show up, feel free to fix it if I messed up

ChainSafeSystems committed:
Update README.md

Commit:
Address comments

ec2 committed:
fix deposit test

ec2 committed:
mock state transition test

Commit:
Clean database before starting dev command

Commit:
A candidate to fix lerna issue in travis

Commit:
Fix linting

Commit:
Change file name and rename createLibP2p

ec2 commented on packages/lodestar/src/sync/initial/initial.ts:
```diff
@@ -0,0 +1,94 @@
+    this.chain.removeListener("processedEpoch", this.sync);
+  }
+
+  private sync = async (chainCheckPoint: Checkpoint) => {
```
sync = async 😂 Nothing wrong here, just think its funny haha

wemeetagain replied:
can we call this fastSync, or somehow differentiate it from a naive sync?
I think we probably will have support for several initial syncing strategies (eg: fast sync, naive sync, light sync, etc)

mpetrunic committed:
rename event

mpetrunic committed:
fix tests

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,83 @@
+
+export function isValidHeaderChain(config: IBeaconConfig, start: BeaconBlockHeader, blocks: BeaconBlock[]): boolean {
+  blocks = blocks.sort((b1, b2) => b1.slot - b2.slot);
+  let previousRoot = hashTreeRoot(start, config.types.BeaconBlockHeader);
```
I think this should be signingRoot. Maybe call the variable parentRoot

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,83 @@
+    if(!previousRoot.equals(block.parentRoot)) {
+      return false;
+    }
+    previousRoot = hashTreeRoot(blockToHeader(config, block), config.types.BeaconBlockHeader);
```
this should be signingRoot, also we don't need the extra `blockToHeader`, we can signingRoot the BeaconBlock directly.

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,83 @@
+    }
+  });
+  if(peersWithHigherFinalizedEpoch.length > 0) {
+    return currentCheckPoint.epoch + 1;
```
I don't understand the + 1 logic.

I think we need to work with a Checkpoint, ie not throw away the root, because the final part of the fast sync strategy is to check the root against the last downloaded block to ensure we've downloaded the chain we think we've downloaded.

mpetrunic replied:
there is no method to fetch non latest finalized checkpoint from peers. I'll extract +1 to constant, basically if you put +2 here it will sync 2 epoch at a time. We can only sync next epoch if there is peer with higher finalized epoch

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,83 @@
+  return true;
+}
+
+export function getTargetEpoch(peers: IReputation[], currentCheckPoint: Checkpoint): Epoch {
```
I think we should change the name here, we use 'target' as a term in FFG votes, makes this kinda confusing.

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,83 @@
+import {blockToHeader} from "../../chain/stateTransition/util";
+import {IReqResp} from "../../network";
+
+export function isValidHeaderChain(config: IBeaconConfig, start: BeaconBlockHeader, blocks: BeaconBlock[]): boolean {
```
I think we should give this name a slightly more descriptive name.
Also, consider adding an end param. (Then we could check the downloaded blocks against a known checkpoint root)
eg:
```typescript
/**
 * Ensure a chain of blocks links properly from start to end
 */
isValidChainOfBlocks(config: IBeaconConfig, start: Hash, end: Hash, blocks: BeaconBlock[]): boolean;
```

mpetrunic replied:
again we don't have end hash at that point, only when we reach highest finalized epoch we can check if our checkpoint matches other peers checkpoint

wemeetagain commented on packages/lodestar/src/sync/utils/sync.ts:
```diff
@@ -0,0 +1,83 @@
+import {IReqResp} from "../../network";
+
+export function isValidHeaderChain(config: IBeaconConfig, start: BeaconBlockHeader, blocks: BeaconBlock[]): boolean {
+  blocks = blocks.sort((b1, b2) => b1.slot - b2.slot);
```
feels sort of strange that this function mutates `blocks` (is relied upon in the calling code)

mpetrunic committed:
fast sync PR changes

mpetrunic committed:
fix ref

mpetrunic committed:
Merge remote-tracking branch 'origin/master' into mpetrunic/initial-sync

# Conflicts:
#	packages/lodestar/src/interop/cli.ts
#	yarn.lock

mpetrunic committed:
fix error

wemeetagain commented on packages/lodestar/src/chain/interface.ts:
```diff
@@ -9,6 +9,7 @@ import {ProgressiveMerkleTree} from "@chainsafe/eth2.0-utils";
 export interface IChainEvents {
   unknownBlockRoot: (root: Hash) => void;
   processedBlock: (block: BeaconBlock) => void;
+  "chain:processedCheckpoint": (checkPoint: Checkpoint) => void;
```
can we stick with `processedCheckpoint` for consistency for now? Maybe tackle 'robust-ifying' our event names separately.

wemeetagain commented on packages/lodestar/src/chain/chain.ts:
```diff
@@ -132,7 +132,8 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
 
-  public async receiveBlock(block: BeaconBlock): Promise<void> {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async receiveBlock(block: BeaconBlock, trusted: boolean = false): Promise<void> {
```
You can string this through now if you merge master.

wemeetagain commented on packages/lodestar/src/sync/initial/fast/index.ts:
```diff
@@ -0,0 +1,98 @@
+import {computeStartSlotOfEpoch} from "@chainsafe/eth2.0-state-transition";
+import {getBlockRange} from "../../utils/blocks";
+
+export class FastSync
```
:+1:

mpetrunic committed:
Merge remote-tracking branch 'origin/master' into mpetrunic/initial-sync

mpetrunic committed:
address PR comments
